File: src/main/java/online/maestoso/soulforged/client/render/ForgedToolItemModel.java
package online.maestoso.soulforged.client.render;

import com.mojang.datafixers.util.Either;
import com.mojang.datafixers.util.Pair;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;

import net.fabricmc.fabric.api.renderer.v1.model.FabricBakedModel;
import net.fabricmc.fabric.api.renderer.v1.render.RenderContext;

import net.minecraft.block.BlockState;

import net.minecraft.client.MinecraftClient;

import net.minecraft.client.render.model.*;
import net.minecraft.client.render.model.json.ItemModelGenerator;
import net.minecraft.client.render.model.json.JsonUnbakedModel;
import net.minecraft.client.render.model.json.ModelOverrideList;
import net.minecraft.client.render.model.json.ModelTransformation;

import net.minecraft.client.texture.Sprite;

import net.minecraft.client.util.SpriteIdentifier;

import net.minecraft.item.ItemStack;

import net.minecraft.nbt.NbtCompound;

import net.minecraft.screen.PlayerScreenHandler;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;

import net.minecraft.world.BlockRenderView;

import online.maestoso.soulforged.Soulforged;

import online.maestoso.soulforged.item.tool.ToolType;
import online.maestoso.soulforged.item.tool.ForgedToolTypes;
import online.maestoso.soulforged.material.Material;
import online.maestoso.soulforged.material.Materials;

import online.maestoso.soulforged.mixin.client.ItemModelGeneratorInvoker;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.function.Function;
import java.util.function.Supplier;

@SuppressWarnings("deprecated")
public class ForgedToolItemModel implements UnbakedModel, BakedModel, FabricBakedModel {
    private static final Identifier ITEM_HANDHELD_MODEL = new Identifier("minecraft:item/handheld");
    private static final HashMap<String, BakedModel> PART_MODELS = new HashMap<>();
    private ModelTransformation transformation;


    public enum ModelToolParts {
        HEAD,
        BINDING,
        HANDLE
    }
    public ForgedToolItemModel() {}
    @Environment(EnvType.CLIENT)
    @Override
    public void emitItemQuads(ItemStack stack, Supplier<Random> randomSupplier, RenderContext context) {
        NbtCompound nbt = stack.getNbt();

        assert nbt != null;
        String type = new Identifier(nbt.getString("sf_tool_type")).getPath();

        String head_mat = new Identifier(nbt.getCompound("sf_head").getString("material")).getPath();
        String binding_mat = new Identifier(nbt.getCompound("sf_binding").getString("material")).getPath();
        String handle_mat = new Identifier(nbt.getCompound("sf_handle").getString("material")).getPath();

        String head_name = modelNameConcatenation(head_mat, type, ModelToolParts.HEAD);
        String binding_name = modelNameConcatenation(binding_mat, type, ModelToolParts.BINDING);
        String handle_name = modelNameConcatenation(handle_mat, type, ModelToolParts.HANDLE);
        BakedModel missingno = MinecraftClient.getInstance().getBakedModelManager().getMissingModel();
        ((FabricBakedModel)PART_MODELS.getOrDefault(head_name, missingno)).emitItemQuads(stack, randomSupplier, context);
        ((FabricBakedModel)PART_MODELS.getOrDefault(binding_name, missingno)).emitItemQuads(stack, randomSupplier, context);
        ((FabricBakedModel)PART_MODELS.getOrDefault(handle_name, missingno)).emitItemQuads(stack, randomSupplier, context);
    }

    @Nullable
    @Override
    public BakedModel bake(ModelLoader loader, Function<SpriteIdentifier, Sprite> textureGetter, ModelBakeSettings rotationContainer, Identifier modelId) {
        JsonUnbakedModel defaultItemModel = (JsonUnbakedModel) loader.getOrLoadModel(ITEM_HANDHELD_MODEL);
        transformation = defaultItemModel.getTransformations();
        for (String mat : Materials.MATERIAL_REGISTRY.getIds().stream().map(Identifier::getPath).toList()) {
            for(String type : ForgedToolTypes.TOOL_TYPES_REGISTRY.getIds().stream().map(Identifier::getPath).toList()) {
                for(ModelToolParts part : ModelToolParts.values()) {
                    String id = modelNameConcatenation(mat, type, part);
                    SpriteIdentifier sprite_id = new SpriteIdentifier(new Identifier("textures/atlas/blocks.png"), new Identifier("soulforged:item/tools/" + type + "/" + mat + "/" + part.toString().toLowerCase()));
                    Sprite sprite = textureGetter.apply(sprite_id);
                    PART_MODELS.put(id, new JsonUnbakedModel(new Identifier("item/handheld"),  ((ItemModelGeneratorInvoker)new ItemModelGenerator()).callAddLayerElements(0, "layer0", sprite), Map.of("layer0", Either.left(sprite_id)), false, null, transformation, List.of()).bake(loader, textureGetter, rotationContainer, modelId));
                }
            }
        }
        return this;
    }

    public static String modelNameConcatenation(String material, String type, ModelToolParts part) {
        return material + "_" + type + "_" + part.toString().toLowerCase();
    }
    @Override
    public Collection<SpriteIdentifier> getTextureDependencies(Function<Identifier, UnbakedModel> unbakedModelGetter, Set<Pair<String, String>> unresolvedTextureReferences) {
        Vector<SpriteIdentifier> ids = new Vector<>();
        for (Material mat : Materials.MATERIAL_REGISTRY.stream().toList()) {
            for(ToolType type : ForgedToolTypes.TOOL_TYPES_REGISTRY.stream().toList()) {
                String mat_name = Objects.requireNonNull(Materials.MATERIAL_REGISTRY.getId(mat)).getPath();
                String type_name = Objects.requireNonNull(ForgedToolTypes.TOOL_TYPES_REGISTRY.getId(type)).getPath();
                SpriteIdentifier head_id = new SpriteIdentifier(new Identifier("textures/atlas/blocks.png"), new Identifier("soulforged:item/tools/" + type_name + "/" + mat_name + "/head"));
                SpriteIdentifier binding_id = new SpriteIdentifier(new Identifier("textures/atlas/blocks.png"), new Identifier("soulforged:item/tools/" + type_name + "/" + mat_name + "/binding"));
                SpriteIdentifier handle_id = new SpriteIdentifier(new Identifier("textures/atlas/blocks.png"), new Identifier("soulforged:item/tools/" + type_name + "/" + mat_name + "/handle"));
                ids.add(head_id);
                ids.add(binding_id);
                ids.add(handle_id);
            }
        }
        return ids;
    }
    // Below is all the stuff I don't care about. It's down here because I don't care about it.
    @Override public Collection<Identifier> getModelDependencies() {
        return Collections.emptyList();
    }

    @Override public List<BakedQuad> getQuads(@Nullable BlockState state, @Nullable Direction face, Random random) {
        Soulforged.LOGGER.fatal("Call to ForgedToolItemModel.getQuads");
        return List.of();
    }
    @Override public boolean useAmbientOcclusion() {
        return false;
    }
    @Override public boolean hasDepth() {
        return false;
    }
    @Override public boolean isSideLit() {
        return false;
    }
    @Override public boolean isBuiltin() {
        return false;
    }
    @Override public Sprite getParticleSprite() {
        return MinecraftClient.getInstance().getSpriteAtlas(PlayerScreenHandler.BLOCK_ATLAS_TEXTURE).apply(new Identifier("block/cobblestone"));
    }
    @Override public ModelTransformation getTransformation() {
        return transformation;
    }
    @Override public ModelOverrideList getOverrides() {
        return ModelOverrideList.EMPTY;
    }
    @Override public boolean isVanillaAdapter() {
        return false;
    }
    @Override public void emitBlockQuads(BlockRenderView blockView, BlockState state, BlockPos pos, Supplier<Random> randomSupplier, RenderContext context) {

    }
}

File: src/main/kotlin/studio/soulforged/soulforged/client/render/ForgedToolItemModel.kt
package studio.soulforged.soulforged.client.render

import com.mojang.datafixers.util.Either
import net.fabricmc.fabric.api.renderer.v1.model.FabricBakedModel
import net.fabricmc.fabric.api.renderer.v1.render.RenderContext
import net.minecraft.block.BlockState
import net.minecraft.client.MinecraftClient
import net.minecraft.client.render.model.*
import net.minecraft.client.render.model.json.ItemModelGenerator
import net.minecraft.client.render.model.json.JsonUnbakedModel
import net.minecraft.client.render.model.json.ModelOverrideList
import net.minecraft.client.render.model.json.ModelTransformation
import net.minecraft.client.texture.Sprite
import net.minecraft.client.util.SpriteIdentifier
import net.minecraft.item.ItemStack
import net.minecraft.screen.PlayerScreenHandler
import net.minecraft.util.Identifier
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.random.RandomGenerator
import net.minecraft.world.BlockRenderView
import org.quiltmc.loader.api.minecraft.ClientOnly
import studio.soulforged.soulforged.client.SoulforgedClient
import studio.soulforged.soulforged.item.tool.ToolInst
import studio.soulforged.soulforged.mixin.client.ItemModelGeneratorInvoker
import java.util.function.Function
import java.util.function.Supplier

class ForgedToolItemModel : UnbakedModel, BakedModel, FabricBakedModel {
    private var transformation: ModelTransformation? = null

    enum class ModelToolParts {
        HEAD, BINDING, HANDLE
    }

    @ClientOnly
    override fun emitItemQuads(stack: ItemStack, randomSupplier: Supplier<RandomGenerator>, context: RenderContext) {
        val tool = ToolInst.ToolInstSerializer.deserialize(stack.nbt!!)
        val type = tool.type.id
        val headName = Identifier(type.namespace, "item/tools/" + type.path.toString() + "_" + ModelToolParts.HEAD.toString().lowercase())
        val bindingName = Identifier(type.namespace, "item/tools/" + type.path.toString() + "_" + ModelToolParts.BINDING.toString().lowercase())
        val handleName = Identifier(type.namespace, "item/tools/" +  type.path.toString() + "_" + ModelToolParts.HANDLE.toString().lowercase())
        val missingno = MinecraftClient.getInstance().bakedModelManager.missingModel
        //TODO recolor instead of loading model from file for all 3 of these
        context.pushTransform(tool.head.mat.transform)
        (PART_MODELS[headName] as FabricBakedModel?)!!.emitItemQuads(
            stack,
            randomSupplier,
            context
        )
        context.popTransform()
        context.pushTransform(tool.binding.mat.transform)
        (PART_MODELS[bindingName] as FabricBakedModel?)!!.emitItemQuads(
            stack,
            randomSupplier,
            context
        )
        context.popTransform()
        context.pushTransform(tool.handle.mat.transform)
        (PART_MODELS[handleName] as FabricBakedModel?)!!.emitItemQuads(
            stack,
            randomSupplier,
            context
        )
        context.popTransform()
    }

    override fun bake(
        modelBaker: ModelBaker?,
        textureGetter: Function<SpriteIdentifier, Sprite>,
        rotationContainer: ModelBakeSettings?,
        modelId: Identifier?
    ): BakedModel {
        val defaultItemModel = modelBaker?.getModel(ITEM_HANDHELD_MODEL) as JsonUnbakedModel
        transformation = defaultItemModel.transformations
        val resources = MinecraftClient.getInstance().resourceManager.findResources("textures/item/tools") { path ->
            return@findResources path.path.endsWith(".png")
        }
        SoulforgedClient.LOGGER.info("Client known tool type parts: ${resources.size}")
        for(id in resources.keys.map {id ->
            Identifier(id.namespace, id.path.substring(9 until id.path.length - 4))
        }) {
            SoulforgedClient.LOGGER.info("Adding model part $id")
                val spriteId = SpriteIdentifier(PlayerScreenHandler.BLOCK_ATLAS_TEXTURE, id)
                val sprite = textureGetter.apply(spriteId)
                PART_MODELS[id] = JsonUnbakedModel(
                    ITEM_HANDHELD_MODEL,
                    (ItemModelGenerator() as ItemModelGeneratorInvoker).callAddLayerElements(0, "layer0", sprite.contents),
                    mapOf("layer0" to Either.left(spriteId)),
                    false,
                    null,
                    transformation,
                    listOf()
                ).bake(modelBaker, textureGetter, rotationContainer, modelId)
        }
        return this
    }

    override fun resolveParents(models: Function<Identifier, UnbakedModel>?) {
        models?.apply(ITEM_HANDHELD_MODEL)
    }

    // Below is all the stuff I don't care about. It's down here because I don't care about it.
    override fun getModelDependencies(): Collection<Identifier> {
        return emptyList()
    }

    override fun getQuads(state: BlockState?, face: Direction?, random: RandomGenerator?): MutableList<BakedQuad> {
        SoulforgedClient.LOGGER.fatal("Call to ForgedToolItemModel#getQuads")
        return mutableListOf()
    }

    override fun useAmbientOcclusion(): Boolean {
        return false
    }

    override fun hasDepth(): Boolean {
        return false
    }

    override fun isSideLit(): Boolean {
        return false
    }

    override fun isBuiltin(): Boolean {
        return false
    }

    override fun getParticleSprite(): Sprite {
        return MinecraftClient.getInstance().getSpriteAtlas(PlayerScreenHandler.BLOCK_ATLAS_TEXTURE)
            .apply(Identifier("block/cobblestone"))
    }

    override fun getTransformation(): ModelTransformation {
        return transformation!!
    }

    override fun getOverrides(): ModelOverrideList {
        return ModelOverrideList.EMPTY
    }

    override fun isVanillaAdapter(): Boolean {
        return false
    }

    override fun emitBlockQuads(
        blockView: BlockRenderView?,
        state: BlockState?,
        pos: BlockPos?,
        randomSupplier: Supplier<RandomGenerator>?,
        context: RenderContext?
    ) {
        TODO("Not yet implemented")
    }

    companion object {
        private val ITEM_HANDHELD_MODEL = Identifier("minecraft:item/handheld")
        val PART_MODELS = HashMap<Identifier, BakedModel?>()
    }
}